### DISCUSSION TITLE
Hajk 3.9 Release Candidate 1

### DISCUSSION THREAD

**Discussion by Hallbergs** [hajkmap/Hajk#1021] [General]
_[303 commits](https://github.com/hajkmap/Hajk/compare/v3.8.0...v3.9.1-rc.1) and changes in 160 files since the 3.8 release – here's Hajk 3.9! This release was brought to you by @Hallbergs  and @jacobwod . Please see the list of commits for all contributors._

_This is the last release that will be built on Material UI 4, all upcoming releases will be built using MUI 5._

# Changelog

### New Draw-plugin (#945) 
The new Draw-plugin presents several new features that were lacking in the old Draw-plugin, some examples:
- Ability to draw arrows.
- Ability to copy drawn objects.
- Ability to move drawn objects a set distance.
- Ability to change the visual appearance of drawn objects.
- Ability to restore removed objects.
   - The removed objects are saved in the `LocalStorage`, so the user must accept `functional cookies` to enable this feature. 
- Ability to save drawings so that the user can continue to draw later.
   - The drawings are saved in the `LocalStorage`, so the user must accept `functional cookies` to enable this feature.

_The new Draw-plugin is not entirely completed, and will get more functionality in the upcoming releases._

![image](https://user-images.githubusercontent.com/43059093/157395258-b737c386-5ee7-4d2d-969c-e17f5d8a67ba.png)

### New InfoDialog plugin to replace Information (#966)
![](https://user-images.githubusercontent.com/110222/148063212-e63c0237-55f1-4958-83f6-c3788f4975b4.png)
- This plugin can be instantiated multiple times per map. Each instance has its own settings, including text, name and icon. 
- The new dialog window use `react-markdown` as renderer. Legacy HTML renderer is supported too.
- Note: there's still no Admin UI interface for configuring this plugin, because this plugins differs so much (as it allows multiple instances per map). Please consult [the plugin's wiki](https://github.com/hajkmap/Hajk/wiki/InfoDialog-plugin) page for a configuration example. 

### Refined Cookie-notice
![image](https://user-images.githubusercontent.com/43059093/157604822-30bd5313-0228-4108-b2b7-10df299f989a.png)
The Cookie-notice (along with its handlers) is now letting the user decide whether they want to accept `required`, `functional` and/or `third-party` cookies. Depending on what the user selects, we are now making sure that we're not storing anything that the user has not accepted to store.

## New features and bug fixes
#345, #806, #840, #855, #856, #862, #865, #872, #878, #887, #909, #917, #935, #939, #941, #945, #964, #966, #968, #970, #971, #980, #990, #991, #998, #999, #1001, #1002, #1006, #1012, #1018, #1019 

## New Contributors
* @johwes made their first contribution in https://github.com/hajkmap/Hajk/pull/1013

**Full Changelog**: https://github.com/hajkmap/Hajk/compare/v3.8.0...v3.9.0-rc.1

---

# Downloads
There are multiple ways to get the latest Hajk up and running, including Docker container.

## Pre-compiled releases
As usual, we have prepared **three different** versions of the pre-compiled solution (scroll down for download links):
1. [Hajk 3.9.0-rc.1 NodeJS edition](https://github.com/hajkmap/Hajk/releases/download/v3.9.0-rc.1/hajk-v3.9.0-rc.1-nodejs.zip)
1. [Hajk 3.9.0-rc.1 Simple edition](https://github.com/hajkmap/Hajk/releases/download/v3.9.0-rc.1/hajk-v3.9.0-rc.1-simple.zip)
1. [Hajk 3.9.0-rc.1 .NET edition](https://github.com/hajkmap/Hajk/releases/download/v3.9.0-rc.1/hajk-v3.9.0-rc.1-dotnet.zip)

## Docker container
The container is available on Docker Hub: https://hub.docker.com/r/hallbergs/hajk

# Installation
Refer to our recently updated [installation guide](https://github.com/hajkmap/Hajk/wiki/Installation-guide-%28for-pre-packaged-releases%29).

<hr /><em>This discussion was created from the release <a href='https://github.com/hajkmap/Hajk/releases/tag/v3.9.0-rc.1'>Hajk 3.9 Release Candidate 1</a>.</em>

**Comment by Hallbergs** [comment]
Reminder: In the new draw plugin, we have to make sure to check how many features the user is about to try and save in the local-storage. As of now, the user can import 50k features from some kml-files and try to store all of them in the local-storage (obviously bad).

**Reply by jacobwod** [reply]
Ouch, I just spent last 4 weeks drawing some 51k+ features in Hajk and you say I can't export them?! 😉 

---

[Here's a site](https://arty.name/localstorage.html) that checks your current browser's LS limits. For my, Safari seems to allow some 5Mb. 

My point is that checking the amount of features might not be the way to go, as it will completely depend on the _complexity_ of each feature. **You can have one mega feature with 5200000 characters (a lot of vertices in that polygon…) and you'll hit the limit.**

Perhaps we can check the length of string prior to saving it to LS?

**Reply by Hallbergs** [reply]
Haha, well you could export them directly to a KML?! :D 

I agree, checking the string-length would be a better solution! Will look at this during the week.

**Reply by Hallbergs** [reply]
Now we're making sure the sketch is not too complex for LS before saving. Might want to tweak the limit in the future, now set to 1e6 characters.

**Comment by lina-sbk-gbg** [comment]
Detected a bug in the FeatureInfo format. Emailed a description to Jacob who says that it is similar to a problem that was discovered and fixed in the migration to MUI5. However, 3.9 is based on MUI4, and it seems that the problem persists here.

**Reply by jacobwod** [reply]
That's correct, I'm certain that it's fixed in `feature/871-upgrade-mui-to-v5`. Perhaps you can ask your organisation to prepare a build for you from that branch and try it out? 

It won't fix the issue in 3.9 of course, but it would be good to know if it's gone in MUI5, and as such will disappear in future releases. 

Also, it would be good to know if other organisations run into this issue in 3.9. If so, perhaps someone can take a closer look on a fix?

**Reply by Hallbergs** [reply]
Just fixed the issue, will be merged to master soon and shipped in 3.9.

**Reply by ingvar-uddevalla** [reply]
I've tested quite a lot of functions in 3.9rc1 and haven't found any bugs yet. Seems promising :-) I've also begun to write a new wiki-page (for the new drawer-plugin). Please add more contents here.

**Reply by lina-sbk-gbg** [reply]
> Just fixed the issue, will be merged to master soon and shipped in 3.9.

I just saw that the color of links in featureinfo is now controlled by MuiTypography-colorInherit. This means that the links does not get a different color. Is that how we want it?

**Reply by jacobwod** [reply]
Correct. Looks like it was introduced as part of https://github.com/hajkmap/Hajk/commit/c66927e605a763dbca2a52a9009b6348404632dc where a change occurred in `new-client/src/utils/customComponentsForReactMarkdown.js`. Perhaps @Hallbergs recalls why?

**Reply by Hallbergs** [reply]
Yes, we did that to get better contrast in dark-mode. (If i recall correctly the idea was that the underlined text would be enough to show the user that they're dealing with a link). Maybe we were wrong?

**Reply by Hallbergs** [reply]
I removed `color="inherit"` on links again so that the links will have `color="primary"`. This will cause bad contrast in dark-mode, but the solution with just using white or black links is not good enough. Will look into this in the next release.

**Comment by WCarlsson** [comment]
Comments from Örebro conserning the new Draw-plugin:
Awesome job! 

When changing opacity to 0% the slider jumps to fully filled (100% mark) and says 10000%. It acts "correctly" and draws the objects with 0% opacity when this happens.
![image](https://user-images.githubusercontent.com/38069156/158408687-42645d21-258e-41f5-9cf2-74473f9a5f00.png)

Expand the possibilities when showing measurments. Lines could show the lengths of the part lenghts, circels the circumference and polygons both of the previous. This could be done by adding more choices in the settings e.g. "Show measurments" and "Show extended measurments" or adding "show part lenghts" "show circumference" etc.

The icons could be changed for the different parts of the module, we suggest the following changes:
- Module icon - old draw icon: Edit icon (mui 4)
- Add feature - old draw icon: Edit icon (mui 4)
- Edit feature - FormatShapes icon (mui 4)
- Move feature - OpenWith icon (mui 4)

Ability to change point size when creating a point and line size when creating an arrow.

Ability to change color opacity on text color.

**Reply by Hallbergs** [reply]
I've sorted all points above except for the extended measurement functionality (i am well under way at sorting that as well though, it would be a really nice addition to the tool). My plan is that the extended measurement-functionality will be added before 3.9 is released.

(The fixes are currently in a separate branch and will be merged soon!)

**Reply by WCarlsson** [reply]
Nice work!
Another comment from the users: Possibility to choose the measurment unit that is displayed. 
I´m thinking the default is the current setting where it changes from m to km or m2 to km2 automatically but there could be a setting to choose to always show one or the other and for area have the option of hectare as well.

**Reply by Hallbergs** [reply]
Absolutely! These settings already exists in the code, but is not displayed for the users yet, will make sure to expose these settings as well.

**Comment by Hallbergs** [comment]
Reminder: Fix #1041 before shipping

**Reply by Hallbergs** [reply]
Fixed

**Comment by emmafrygne-vbg** [comment]
VARBERG
Nice work with the new Draw-plugin! We think it looks good and works good. Its good that text is visible when moving it and not only the dot. Its good that you can regret erased objects. Its good that you still can draw by hand when pressing shift, maybe it can be written somewhere that thats possible? Great that you can save now. How long is it saved?

We have some requests: 
- It would be good if black and white can be included in the panel 
![image](https://user-images.githubusercontent.com/59825341/158987872-f7423d70-c548-4760-87fe-bb4cb5842bd8.png)
- Show measure should be more visible, maybe a separate bottom instead of under settings 
- Would have been good to be able to draw a circle and choose radius/diameter

Another comment is about kartinteraktioner in admin. If the settings are not in config its default that you cant pan the map etc. See picture below, I hope you understand what I mean :) 
![image](https://user-images.githubusercontent.com/59825341/158998208-11afb5f7-7a36-4c1f-99da-c0d67aacfd28.png)

**Reply by Hallbergs** [reply]
Regarding saved sketches: They are saved in `local-storage`, so they don't have an expiration date. As long as the user doesn't clear their storage, they will be persisted. (If you change the domain for the map they will disappear obviously).

Regarding the other points they are all valid, will try to get them all fixed!

**Reply by jacobwod** [reply]
> Another comment is about kartinteraktioner in admin. If the settings are not in config its default that you cant pan the map etc. See picture below, I hope you understand what I mean :) ![image](https://user-images.githubusercontent.com/59825341/158998208-11afb5f7-7a36-4c1f-99da-c0d67aacfd28.png)

I was about to get this fixed, but **I couldn't reproduce the issue**. Please see the following video. It uses `simpleMapConfig.json` and the only thing I modify there is the `onFocusOnly` key on the `map` object. 

In the first case, `onFocusOnly` option is set to `false`. The panning interaction works all the time.

In the second case, I enable the `onFocusOnly` (`true`), so the must first gain focus, before it can be panned. 

Everything works as expected so far.

Finally, I remove the option from config entirely (it gets the value of `undefined`). In this case the resulting behavior is as expected (and as intended): the map can be panned even without initial focus.

https://user-images.githubusercontent.com/110222/159237971-08626479-cf4a-4b6e-b328-5fdd9bf2c92d.mov

Are you sure that this occurs when `onFocusOnly` is `undefined`?

**Reply by emmafrygne-vbg** [reply]
Ok thanks, it works now when we have everything in config.

**Reply by jacobwod** [reply]
Still, there might be something buggy in Admin, where the keys or its' values aren't set as expected. Should investigate further.

**Comment by Hallbergs** [comment]
I've decided to leave some of the "fixes" in the new-draw-plugin to the next release. The points i've decided not to fix right now (since they feel too big to change in a release candidate) are:

- Color-panel, allow admins to select colors.
- Move the "show measurements" to a more visible place.
- Draw a circle with fixed radius.
- Select size on points and arrows.
- Select opacity on text-colors.
- Show length on each part of a line-string.

The rest is taken care of. I think we should create an issue with further draw-improvements after 3.9 is released.

**Reply by jacobwod** [reply]
I second that decision: we should not introduce new features in the RC-stage.

**Comment by Hallbergs** [comment]
I found an issue regarding the new printing functionality (where we are making sure tiled WMS are requested with the correct DPI). It seems as i reset the layers back to their original DPI a bit to early, leading to a possibility that the resulting print contained mixed tiles (with some tiles with the printing DPI and some with the map-dpi). 

It's fixed in master now though.
